### PR TITLE
GPU passthru for NVIDIA cards bsc#1169904

### PR DESCRIPTION
**Pull request by tbazant**
### Description

This new appendix describes specific steps to configure pass-through for NVIDIA Tesla graphic cards
Rendered PDF is attached 
[app-gpu-passthru_color_en.pdf](https://github.com/SUSE/doc-sle/files/6653627/app-gpu-passthru_color_en.pdf)

### Are there any relevant issues/feature requests?

* bsc#1169904
* jsc#DOCTEAM-98

### To which product versions do the changes apply?

The first column is to be filled by the requester, the second column is to be filled by the doc team after the changes have been backported to the maintenance branches.

| Code 15     | Backport Done
| ----------  | ----------
| [x] 15 SP4  |*(no backport necessary)*
| [x] 15 SP3  | [ ]
| [ ] 15 SP2  | [ ]
| [ ] 15 SP1  | [ ]
| [ ] 15 SP0  | [ ]

| Code 12     | Backport Done
| ----------  | ----------
| [ ] 12 SP5  | [ ]
| [ ] 12 SP4  | [ ]
| [ ] 12 SP3  | [ ]
| [ ] 12 SP2  | [ ]

### CODE REVIEW

File: xml/gpu_passthru.xml [hunk 1]
```diff
@@ -0,0 +1,503 @@
...
+  <!ENTITY % entities SYSTEM "entity-decl.ent">
+    %entities;
+]>
+<appendix xmlns="http://docbook.org/ns/docbook" xmlns:xi="http://www.w3.org/2001/XInclude" xmlns:xlink="http://www.w3.org/1999/xlink" version="5.0" xml:id="app-gpu-passthru">
+ <title>Configuring &gpuback; for &nvidia; cards</title>
+ <sect1 xml:id="gpu-passthru-intro">
+  <title>Introduction</title>
+
+  <para>
+   This article describes how to assign an &nvidia; GPU graphical card on the
```
Comment (lproven) [inline xml/gpu_passthru.xml:13]:
```suggestion
   This article describes how to assign an &nvidia; GPU graphics card on the
```

File: xml/gpu_passthru.xml [hunk 2]
```diff
@@ -0,0 +1,503 @@
...
+   </listitem>
+   <listitem>
+    <para>
+     The host system needs to SLES 12 SP3 or newer.
+    </para>
+   </listitem>
+   <listitem>
+    <para>
+     This article deals with a set of instructions based on V100/T1000 &nvidia;
+     cards, and is meant for GPU compute purpose only.
```
Comment (lproven) [inline xml/gpu_passthru.xml:34]:
```suggestion
     cards, and is meant for GPU computation purposes only.
```

Comment (lproven) [reply xml/gpu_passthru.xml:34]: Possibly GPGPU? (General Purpose Graphics Processing Unit)? This is the normal term for running computational tasks on a GPU.

File: xml/gpu_passthru.xml [hunk 3]
```diff
@@ -0,0 +1,503 @@
...
+<screen>&prompt.user;dmesg |  grep -e DMAR -e IOMMU</screen>
+    </step>
+   </procedure>
+  </sect2>
+
+  <sect2 xml:id="gpu-passthru-blacklist-nouveau">
+   <title>Blacklist the Nouveau driver</title>
+   <para>
+    Because we want to assign the &nvidia; card to a VM guest, we need to avoid
+    using the card by the host opensource driver. Edit the file
```
Comment (lproven) [inline xml/gpu_passthru.xml:147]:
```suggestion
    use of the card by the host open source &nvidia; driver. Edit the file
```

File: xml/gpu_passthru.xml [hunk 4]
```diff
@@ -0,0 +1,503 @@
...
+     <para>
+      Create the file <filename>vfio.conf</filename> in the
+      <filename>/etc/modprobe.d/</filename> directory with the following
+      content:
+     </para>
+<screen>options vfio-pci ids=10de:1db4</screen>
+     <note>
+      <para>
+       Double check that your card does not need an extra <option>ids=</option>
+       parameter. For example, for Public Nvidia Card you need to also add the
```
Comment (lproven) [inline xml/gpu_passthru.xml:178]: I do not understand the phrase "Public Nvidia Card", and it gets zero hits on Google. What does this mean? If we can rephrase it to something clearer, we should probably be using the nVidia entity, too.

File: xml/gpu_passthru.xml [hunk 5]
```diff
@@ -0,0 +1,503 @@
...
+    </para>
+<screen>&prompt.sudo;modprobe vfio-pci</screen>
+   </sect3>
+  </sect2>
+
+  <sect2 xml:id="gpu-passthru-disable-msr">
+   <title>Disable MSR for &mswin; guests</title>
+   <para>
+    For &mswin; guests, we recommend disabling MSR (model-specific register) to
+    avoid crash of the guest. Create the file
```
Comment (lproven) [inline xml/gpu_passthru.xml:240]:
```suggestion
    avoid the guest crashing. Create the file
```

File: xml/gpu_passthru.xml [hunk 6]
```diff
@@ -0,0 +1,503 @@
...
+    <filename>/etc/modprobe.d/kvm.conf</filename> and add the following
+    content:
+   </para>
+<screen>options kvm ignore_msrs=1</screen>
+  </sect2>
+
+  <sect2 xml:id="gpu-passthru-ovmf">
+   <title>Install and enable UEFI firmware</title>
+   <para>
+    For proper &gpuback; functionality, the host need to boot using the UEFI
```
Comment (lproven) [inline xml/gpu_passthru.xml:250]:
```suggestion
    For proper &gpuback; functionality, the host needs to boot using the UEFI
```

File: xml/gpu_passthru.xml [hunk 7]
```diff
@@ -0,0 +1,503 @@
...
+    firmware.
```
Comment (lproven) [inline xml/gpu_passthru.xml:251]:
```suggestion
    firmware (that is, not using a legacy-style BIOS boot sequence).
```

Comment (lproven) [reply xml/gpu_passthru.xml:251]: This suggestion is partly to reduce ambiguity -- it is not possible to choose the type of firmware in a computer, but it is possible to choose what boot mode the OS uses.

File: xml/gpu_passthru.xml [hunk 8]
```diff
@@ -0,0 +1,503 @@
...
+   </para>
+   <procedure>
+    <step>
+     <para>
+      Install the <package>qemu-ovmf</package> package that includes UEFI
```
Comment (lproven) [inline xml/gpu_passthru.xml:256]:
```suggestion
      Install the <package>qemu-ovmf</package> package which includes UEFI
```

File: xml/gpu_passthru.xml [hunk 9]
```diff
@@ -0,0 +1,503 @@
...
+   </para>
+<screen>&prompt.sudo;shutdown -r now</screen>
+  </sect2>
+ </sect1>
+ <sect1 xml:id="gpu-passthru-guest">
+  <title>Configuring the guest</title>
+
+  <para>
+   This section describes how to configure the guest virtual machine so that it
+   can use the host's GPU &nvidia; graphic card. Use &vmm; or
```
Comment (lproven) [inline xml/gpu_passthru.xml:309]:
```suggestion
   can use the host's &nvidia; GPU. Use &vmm; or
```

Comment (lproven) [reply xml/gpu_passthru.xml:309]: As the earlier instructions in the chapter made clear, the GPU for pass-through is **not** to be the default graphics card, i.e. the card displaying the host's OS screen. It must be a dedicated GPU.

AIUI the Tesla series do not actually have monitor outputs and so cannot be used to display graphics. They are dedicated GPGPUs and not graphics cards.

File: xml/gpu_passthru.xml [hunk 10]
```diff
@@ -0,0 +1,503 @@
...
+     <step>
+      <para>
+       Because the installer needs to compile the &nvidia; driver modules,
+       install <package>gcc-c++</package> and <package>kernel-devel</package>
+       packages.
+      </para>
+     </step>
+     <step>
+      <para>
+       Disable secure boot on the guest because &nvidia; driver modules are not
```
Comment (lproven) [inline xml/gpu_passthru.xml:404]:
```suggestion
       Disable Secure Boot on the guest, because the &nvidia; driver modules are not
```

File: xml/gpu_passthru.xml [hunk 11]
```diff
@@ -0,0 +1,503 @@
...
+       signed. On &suse; distributions, you can use &yast; &grub; module to
+       disable secure boot. Find more information in
```
Comment (lproven) [inline xml/gpu_passthru.xml:406]:
```suggestion
       disable Secure Boot. Find more information in
```

**Review by lproven — CHANGES_REQUESTED** [review]

File: xml/gpu_passthru.xml [hunk 12]
```diff
@@ -0,0 +1,503 @@
...
+      Create the file <filename>vfio.conf</filename> in the
+      <filename>/etc/modprobe.d/</filename> directory with the following
+      content:
+     </para>
+<screen>options vfio-pci ids=10de:1db4</screen>
+     <note>
+      <para>
+       Double check that your card does not need an extra <option>ids=</option>
+       parameter. For example, for Public Nvidia Card you need to also add the
+       audio device in the list, otherwise you will not be able to use the
```
Comment (tbazant) [inline xml/gpu_passthru.xml:179]:
```suggestion
       Verify that your card does not need an extra <option>ids=</option>
       parameter. Some cards need to also add the
       audio device in the list, otherwise you will not be able to use the
```

**Review by tbazant — COMMENTED** [review]
More genralized one sentence

File: xml/gpu_passthru.xml [hunk 13]
```diff
@@ -0,0 +1,503 @@
...
+
+  <itemizedlist>
+   <listitem>
+    <para>
+     GPU pass-through is supported on the &x86-64; architecture only.
+    </para>
+   </listitem>
+   <listitem>
+    <para>
+     The host system needs to SLES 12 SP3 or newer.
```
Comment (lproven) [inline xml/gpu_passthru.xml:28]:
```suggestion
     The host operating system needs to be SLES 12 SP3 or newer.
```

File: xml/gpu_passthru.xml [hunk 14]
```diff
@@ -0,0 +1,503 @@
...
+ </sect1>
+ <sect1 xml:id="gpu-passthru-host">
+  <title>Configuring the host</title>
+
+  <sect2 xml:id="gpu-passthru-verify-host">
+   <title>Verify the host environment</title>
+   <procedure>
+    <step>
+     <para>
+      Verify that the host system is SLES 12 SP3 or newer:
```
Comment (lproven) [inline xml/gpu_passthru.xml:60]:
```suggestion
      Verify that the host operating system is SLES 12 SP3 or newer:
```

Comment (lproven) [reply xml/gpu_passthru.xml:60]: I am clarifying this because in context I would assume "host system" to mean the server hardware.

File: xml/gpu_passthru.xml [hunk 15]
```diff
@@ -0,0 +1,503 @@
...
+     </para>
+<screen>
+&prompt.user;cat /etc/issue
+Welcome to SUSE Linux Enterprise Server 15  (x86_64) - Kernel \r (\l).
+</screen>
+    </step>
+    <step>
+     <para>
+      Verify that the host supports <xref linkend="gloss-vt-acronym-vtd"/>
+      technology and that it is already enabled in the BIOS:
```
Comment (lproven) [inline xml/gpu_passthru.xml:70]:
```suggestion
      technology and that it is already enabled in the firmware settings:
```

Comment (lproven) [reply xml/gpu_passthru.xml:70]: (Since later we specifically say that it needs to be UEFI, **not** a BIOS.)

File: xml/gpu_passthru.xml [hunk 16]
```diff
@@ -0,0 +1,503 @@
...
+     </para>
+<screen>
+&prompt.user;dmesg | grep -e "Directed I/O"
+[   12.819760] DMAR: Intel(R) Virtualization Technology for Directed I/O
+</screen>
+     <para>
+      If VT-d is not enabled in the BIOS, enable it and reboot the host.
```
Comment (lproven) [inline xml/gpu_passthru.xml:77]:
```suggestion
      If VT-d is not enabled in the firmware, enable it and reboot the host.
```

File: xml/gpu_passthru.xml [hunk 17]
```diff
@@ -0,0 +1,503 @@
...
+   </procedure>
+  </sect2>
+
+  <sect2 xml:id="gpu-passthru-blacklist-nouveau">
+   <title>Blacklist the Nouveau driver</title>
+   <para>
+    Because we want to assign the &nvidia; card to a VM guest, we need to avoid
+    use of the card by the host open source &nvidia; driver. Edit the file
+    <filename>/etc/modprobe.d/50-blacklist.conf</filename> and append the
+    following line to its end:
```
Comment (lproven) [inline xml/gpu_passthru.xml:149]:
```suggestion
    Because we want to assign the &nvidia; card to a VM guest, we need to avoid
    use of the card by the host OS's built-in driver for &nvidia; GPUs. The open
    source &nvidia; driver is called <literal>nouveau</literal>. Edit the file 
    <filename>/etc/modprobe.d/50-blacklist.conf</filename> and append the
    following line to its end:
```

Comment (lproven) [reply xml/gpu_passthru.xml:149]: Since there are multiple nVidia drivers, the specification of "the host open source driver" seemed odd to me. Why just the open-source one? Why not disable any/all drivers? The next line makes it clear -- it's not disabling the driver completely, it's making sure that Nouveau doesn't load so that the proprietary driver **can** grab the card. As such I thought it was worth clarifying which driver we are talking about.

File: xml/gpu_passthru.xml [hunk 18]
```diff
@@ -0,0 +1,503 @@
...
+      <filename>/etc/modprobe.d/</filename> directory with the following
+      content:
+     </para>
+<screen>options vfio-pci ids=10de:1db4</screen>
+     <note>
+      <para>
+       Verify that your card does not need an extra <option>ids=</option>
+       parameter. Some cards need to also add the
+       audio device in the list, otherwise you will not be able to use the
+       card.
```
Comment (lproven) [inline xml/gpu_passthru.xml:180]:
```suggestion
       Verify that your card does not need an extra <option>ids=</option>
       parameter. For some cards, you must specify the audio device too,
       so that device's ID must also be added to the list, otherwise you
       will not be able to use the card.
```

File: xml/gpu_passthru.xml [hunk 19]
```diff
@@ -0,0 +1,503 @@
...
+    <filename>/etc/modprobe.d/kvm.conf</filename> and add the following
+    content:
+   </para>
+<screen>options kvm ignore_msrs=1</screen>
+  </sect2>
+
+  <sect2 xml:id="gpu-passthru-ovmf">
+   <title>Install and enable UEFI firmware</title>
+   <para>
+    For proper &gpuback; functionality, the host needs to boot using the UEFI
```
Comment (lproven) [inline xml/gpu_passthru.xml:250]:
```suggestion
    For proper &gpuback; functionality, the host needs to boot using UEFI
```

File: xml/gpu_passthru.xml [hunk 20]
```diff
@@ -0,0 +1,503 @@
...
+   </para>
+   <itemizedlist>
+    <listitem>
+     <para>
+      Use Q35 chipset if possible.
+     </para>
+    </listitem>
+    <listitem>
+     <para>
+      Install the guest VM by using the UEFI firmware.
```
Comment (lproven) [inline xml/gpu_passthru.xml:328]:
```suggestion
      Install the guest VM using UEFI firmware.
```

File: xml/gpu_passthru.xml [hunk 21]
```diff
@@ -0,0 +1,503 @@
...
+       Because the installer needs to compile the &nvidia; driver modules,
+       install <package>gcc-c++</package> and <package>kernel-devel</package>
+       packages.
+      </para>
+     </step>
+     <step>
+      <para>
+       Disable Secure Boot on the guest, because the &nvidia; driver modules are not
+       signed. On &suse; distributions, you can use &yast; &grub; module to
+       disable Secure Boot. Find more information in
```
Comment (lproven) [inline xml/gpu_passthru.xml:406]:
```suggestion
       Disable Secure Boot on the guest, because &nvidia;'s driver modules are
       unsigned. On &suse; distributions, you can use the &yast; &grub; module
       to disable Secure Boot. Find more information in
```

File: xml/gpu_passthru.xml [hunk 22]
```diff
@@ -0,0 +1,503 @@
...
+       <xref linkend="sec-uefi-secboot-sle"/>.
+      </para>
+     </step>
+     <step>
+      <para>
+       Download the driver installation script from
+       <link xlink:href="https://www.nvidia.com/Download/index.aspx?lang=en-us"/>,
+       make it executable, and execute it, and complete the driver
```
Comment (lproven) [inline xml/gpu_passthru.xml:414]:
```suggestion
       make it executable, and run it to complete the driver
```

File: xml/gpu_passthru.xml [hunk 23]
```diff
@@ -0,0 +1,503 @@
...
+     </step>
+    </procedure>
+    <note>
+     <title>Display issues</title>
+     <para>
+      After you have installed the &nvidia; drivers, the &vmm; display will
+      lose connection. To access the guest VM, you need to login via
+      <command>ssh</command>, change to console interface, or install a
+      dedicated VNC server in the guest. To avoid a flickering screen, stop and
+      disable the display manager:
```
Comment (lproven) [inline xml/gpu_passthru.xml:437]:
```suggestion
      After you have installed the &nvidia; drivers, the &vmm; display will
      lose its connection to the guest OS. To access the guest VM, you
      must either login via <command>ssh</command>, change to the
      console interface, or install a dedicated VNC server in the guest.
      To avoid a flickering screen, stop and disable the display manager:
```

**Review by lproven — CHANGES_REQUESTED** [review]
This is looking much clearer now... but because I can understand more of the steps better, I am afraid I have more changes. Sorry. :-(

File: xml/gpu_passthru.xml [hunk 24]
```diff
@@ -0,0 +1,504 @@
...
+      </para>
+<screen>&prompt.sudo;shutdown -r now</screen>
+     </step>
+    </procedure>
+    <procedure>
+     <title>Generic installer</title>
+     <step>
+      <para>
+       Because the installer needs to compile the &nvidia; driver modules,
+       install <package>gcc-c++</package> and <package>kernel-devel</package>
```
Comment (lproven) [inline xml/gpu_passthru.xml:399]:
```suggestion
       install the <package>gcc-c++</package> and <package>kernel-devel</package>
```

**Review by lproven — CHANGES_REQUESTED** [review]
One missing article & then we are good to go.

**Comment by liayan** [comment]
@tbazant, could you attach a pdf file here as you did for others? Thanks.

**Comment by tbazant** [comment]
> @tbazant, could you attach a pdf file here as you did for others? Thanks.
@liayan done, the PDF is at https://github.com/SUSE/doc-sle/files/6653627/app-gpu-passthru_color_en.pdf

**Review by lproven — APPROVED** [review]

**Comment by liayan** [comment]
Nice work! All look good to me.
